pkg/tlsutil/tls.go

```diff
@@ -111,3 +119,135 @@ type CertGenerator interface {
 	//   ca.key: ..
 	GenerateCert(cr runtime.Object, service *v1.Service, config *CertConfig) (*v1.Secret, *v1.ConfigMap, *v1.Secret, error)
 }
+
+const (
+	// TLSPrivateCAKeyKey is the key for the private CA key field.
+	TLSPrivateCAKeyKey = "ca.key"
+	// TLSCertKey is the key for tls CA certificates.
+	TLSCACertKey = "ca.crt"
+)
+
+type SDKCertGenerator struct {
+	// appKeyAndCertMap holds the application key and cert for a given cr + config.CertName.
+	appKeyAndCertMap sync.Map
+	// caKeyAndCertMap holds the CA key and cert for a given cr.
+	caKeyAndCertMap sync.Map
+}
+
+type keyAndCert struct {
+	key  *rsa.PrivateKey
+	cert *x509.Certificate
+}
+
+func (scg *SDKCertGenerator) GenerateCert(cr runtime.Object, service *v1.Service, config *CertConfig) (*v1.Secret, *v1.ConfigMap, *v1.Secret, error) {
+	if err := verifyConfig(config); err != nil {
+		return nil, nil, nil, err
+	}
+
+	n, k, ns, err := toKindNameNamespace(cr)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	asn := toAppSecretName(k, n, config.CertName)
+	cascn := toCASecretAndConfigMapName(k, n)
+	appkcv, hasAppKeyAndCert := scg.appKeyAndCertMap.Load(asn + "-" + ns)
+	cakcv, hasCAKeyAndCert := scg.caKeyAndCertMap.Load(cascn + "-" + ns)
+	// TODO: handle passed in CA
+	if hasAppKeyAndCert && hasCAKeyAndCert {
+		s := toTLSSecret(appkcv.(*keyAndCert), asn, ns)
```

Comment:
If the AppKeyAndCert and CAKeyAndCert are found then, the secrets and config map should be in the cluster. If that is true, I would expect that the secrets and config maps that are returned are the actual values from the cluster. 

The reason I think this could be important is if for whatever reason a user wants to update those values, they will not be able to do, do to resource version mismatch I believe.

Reply:
@shawn-hurley that makes sense. Now if we want to maintain the cache with your suggestion, it can be tricky because once user updates the secret, the one in the cache is going to be outdated. Maintaining cache consistency is hard and error prone. Alternatively, CertGenerator always queries Kubernetes cluster to check for the existence of the Application Key and Cert and CA Key and Cert. However, this approach can leads to potential performance issue. I am fine going to the latter approach for the sake of simplicity. We can tackle the cache optimization later. Let's aim for correctness first.  

cc/ @shawn-hurley and @hasbro17